yyyymmdd/range.py

```diff
@@ -351,9 +382,174 @@ def stop_max_inv(x, y):
+            (other._ordrange.start <= self._ordrange.start < other._ordrange.stop)
+            or (self._ordrange.start <= other._ordrange.start < self._ordrange.stop)
+        )
+
+    def underlaps(self, other):
+        """
+        Finds the underlap of this DateRange and the given DateRange other.
+
+        The X and Y underlap if X and Y overlap or are adjacent.
+        """
+        if self.step != 1 or other.step != 1:
+            raise ValueError(
+                "Mereological methods are undefined for ranges with steps != 1: %r & %r"
+                % (self, other)
+            )
+        return self.is_adjacent(other) or self.overlaps(other)
+
+    def __or__(self, other):
+        return self.union(other)
+
+    def union(self, other):
+        """
+        Computes the union of this DateRange and the given DateRange other.
+
+        The union of X and Y is only defined if X and Y underlap.
+        """
+        if self.step != 1 or other.step != 1:
+            raise ValueError(
+                "Mereological methods are undefined for ranges with steps != 1: %r & %r"
+                % (self, other)
+            )
+        if not self.underlaps(other):
+            raise ValueError(
+                "DateRange union is undefined for ranges that are not adjacent and do not overlap: %r & %r"
+                % (self, other)
+            )
+        return DateRange(
+            self._o2d(min(self._ordrange.start, other._ordrange.start)),
+            self._o2d(max(self._ordrange.stop, other._ordrange.stop)),
+        )
+
+    def __sub__(self, other):
+        return self.difference(other)
+
+    def difference(self, other):
+        """
+        Computes the dates from this DateRange less the dates in the given
+        DateRange other.
+
+        This method returns a list, because the difference can result
+        up to two contiguous date ranges.
+        """
+        if self.step != 1 or other.step != 1:
+            raise ValueError(
+                "Mereological methods are undefined for ranges with steps != 1: %r & %r"
+                % (self, other)
+            )
+        # ***A***
+        #   ***B***
+        # gives:
+        # **
+        #
+        #   ***A***
+        # ***B***
+        # gives:
+        #        **
+        #
+        # *****A*****
+        #   **B**
+        # gives:
+        # **     ****
+        if self.overlaps(other):
+            part1 = DateRange(
+                self._o2d(self._ordrange.start), self._o2d(other._ordrange.start)
+            )
+            part2 = DateRange(
+                self._o2d(other._ordrange.stop), self._o2d(self._ordrange.stop)
+            )
+            if bool(part1):
+                if bool(part2):
+                    return [part1, part2]
+                else:
+                    return [part1]
+            else:
+                if bool(part2):
+                    return [part2]
+                return [DateRange.empty()]
```

Comment:
Why return a list with an empty range? An empty list is just as good.
```suggestion
                return []
```